index.js

```diff
@@ -21,30 +20,64 @@ const TranslationReducer = require('./lib/broccoli/translation-reducer');
 
 module.exports = {
   name: 'ember-intl',
-  addonOptions: null,
+  opts: null,
   isLocalizationFramework: true,
 
   included() {
     this._super.included.apply(this, arguments);
 
-    let app = this._findHost();
-    this.app = app;
-    this.addonOptions = this.intlConfig(app.env);
+    this.app = this._findHost();
+    this.opts = this.getOptions(this.app.env);
+    this.locales = this.findLocales();
+    this.translationTree = this._buildTranslationTree();
+  },
 
-    let inputPath = this.addonOptions.inputPath || 'translations';
-    this.hasTranslationDir = fs.existsSync(path.join(app.project.root, inputPath));
-    this.projectLocales = this.findLocales();
+  _buildTranslationTree() {
+    const projectTranslations = path.join(this.project.root, this.opts.inputPath || 'translations');
+    const translationTrees = [];
 
-    let projectTranslations = new WatchedDir(inputPath);
+    this._processAddons(this.project.addons, translationTrees);
 
-    let addonTranslations = this.findIntlAddons().map(function(addon) {
-      return funnel(addon.path, {
-        srcDir: addon.translationPath,
-        destDir: `__addon__${addon.name}`
-      });
-    }, this);
+    if (fs.existsSync(projectTranslations)) {
+      translationTrees.push(this.treeGenerator(projectTranslations));
+    }
+
+    return funnel(
+      mergeTrees(translationTrees, {
+        overwrite: true
+      }),
+      {
+        include: ['**/*.yaml', '**/*.yml', '**/*.json']
+      }
+    );
+  },
+
+  _processAddons(addons, translationTrees) {
+    addons.forEach(addon => this._processAddon(addon, translationTrees));
+  },
 
-    this.translationTree = this.mergeTranslationTrees(projectTranslations, addonTranslations);
+  /**
+   * NOTE: `_processAddon` was heavily adapted off the work of ember-cli-fastboot implementation.
+   */
+  _processAddon(addon, translationTrees) {
+    const addonTranslationPath = path.join(addon.root, 'translations');
+    let addonGeneratedTree;
+
+    if (fs.existsSync(addonTranslationPath)) {
+      addonGeneratedTree = this.treeGenerator(addonTranslationPath);
+    }
+
+    if (addon.treeForTranslations) {
+      let additionalTranslationTree = addon.treeForTranslations(addonGeneratedTree);
+
+      if (additionalTranslationTree) {
+        translationTrees.push(funnel(additionalTranslationTree, { destDir: addon.name }));
+      }
+    } else if (addonGeneratedTree !== undefined) {
+      translationTrees.push(funnel(addonGeneratedTree, { destDir: addon.name }));
```

Comment:
Translations from addons are namespaced within a folder based on the addon's name.  This is so that the app and the addon translations do not collide.  In the case of a rare case of collision of a file (i.e., the app creates a folder with the name of an addon) then the app will win.